### ISSUE TITLE
Inability to specify binary caveat keys for 3rd party breaks 3rd party caveat validation

### ISSUE THREAD

**Issue by jaxley** [nitram509/jmacaroons#11]
The API lets you specify either a String as a root key or a byte array for root keys.  If you specify a string, then internally it uses a key derivation algorithm to derive a key from that (presumably assuming someone using the String-based APIs is providing a low-quality password that needs to be expanded into a derived key?)

However, there is no such byte array option for third party caveats.  You can only specify a key by a String, which results in a derived value that you don't control.

Was tearing my hair out trying to understand why the validation was failing until I realized that the key used as byte array in one instance results in a signature different than the one expected because the string key is changed to a derived key so they'll never match.

Unit test to demonstrate the error (requires java language level of 1.8+):

```java
  @Test
  public void using_non_string_key_bytes_for_3rd_party__should_succeed() {
    byte[] root_key = keyGen();
    byte[] caveat_key = keyGen();
    String caveat_id = "caveat";
    String macaroon_id = "123456";
    Macaroon M = new MacaroonsBuilder("some-service", root_key, macaroon_id)
        .add_third_party_caveat("authN", new String(caveat_key), caveat_id)
        .getMacaroon();

    assertThat(new MacaroonsVerifier(M)
        .isValid(root_key))
        .describedAs("Original should not be valid without discharge macaroon")
        .isFalse();

    // create discharge macaroon using caveat key as bytes
    Macaroon D = new MacaroonsBuilder("authN", caveat_key, caveat_id)
        .getMacaroon();

    assertThat(new MacaroonsVerifier(D)
        .isValid(caveat_key))
        .describedAs("Discharge macaroon should be valid")
        .isTrue();

    // prepare discharge macaroon by binding to the original macaroon
    Macaroon bound = new MacaroonsBuilder(M)
        .prepare_for_request(D)
        .getMacaroon();

    assertThat(new MacaroonsVerifier(bound)
        .isValid(caveat_key))
        .describedAs("Bound discharge macaroon should not be considered valid on its own")
        .isFalse();

    assertThat(new MacaroonsVerifier(M)
        .satisfy3rdParty(bound)
        .isValid(root_key))
        .describedAs("Original should be valid with third party caveat bound")
        .isTrue();
  }

  byte[] keyGen() {
    try {
      byte[] keyBytes = new byte[32];
      SecureRandom prng = SecureRandom.getInstanceStrong();
      prng.nextBytes(keyBytes);
      return keyBytes;
    }
    catch (NoSuchAlgorithmException e) {
      e.printStackTrace();
    }
    return null;
  }
```

**Comment by jaxley** [comment]
Looking closer at the reference implementation, it seems that this is more or less a Java port of that implementation, APIs and all.  That implementation always does a derived key for the caveat key.  The difference in implementations is that macaroon keys are *always* derived no matter if they start life as an array of chars from a string or an array of chars as key bytes.  

So, that would be an alternative fix than adding new APIs to the library.

**Comment by jaxley** [comment]
Code is currently using UTF-8 when converting string to bytes, which is not reversible.  This makes mixing byte arrays and strings impossible.  Playing around with a fix.  It looks like various implementations are taking different approaches to how the keys are derived in these cases.  I can only imagine macaroons from one library are not compatible...

```
// This will succeed
    assertThat(new String(caveat_key, Charset.forName("ISO-8859-1")).getBytes(Charset.forName("ISO-8859-1")))
        .describedAs("Bytes from string should match original byte array when converted using iso-8859-1")
        .isEqualTo(caveat_key);

// this will FAIL
    assertThat(new String(caveat_key, MacaroonsConstants.IDENTIFIER_CHARSET).getBytes(MacaroonsConstants.IDENTIFIER_CHARSET))
        .describedAs("Bytes from string should match original byte array")
        .isEqualTo(caveat_key);
```

**Comment by jaxley** [comment]
I made a different fix that follows the other macaroon implementation approaches of always ensuring that keys provided are used to generate derived fixed-length keys.

**Comment by nitram509** [comment]
Solved, with merging PR #13 - thank you Jason, for this contribution.
